### PR TITLE
Implement support CFD for dpnp.isclose and dpnp.allclose functions

### PR DESCRIPTION
**Pull request by npolina4**
Implement support CFD for dpnp.isclose and dpnp.allclose functions.
Cleaned up legacy logic implementation from the backend.
Added test.

Performance table:
``` python
np.isclose(a, b)
```
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>

<div style='direction:ltr'>

  | float 32 |   |   | float 64 |  
-- | -- | -- | -- | -- | --
Size | numpy | dpnp cpu | dpnp gpu | numpy | dpnp cpu
10.000 | 0.0001 | 0.001 | 0.001 | 0.0001 | 0.001
100.000 | 0.0009 | 0.004 | 0.002 | 0.0007 | 0.003
1.000.000 | 0.0035 | 0.0033 | 0.002 | 0.0061 | 0.0055
10.000.000 | 0.04 | ~0.03~ 0.02 | ~0.06~ 0.03 | 0.06 | ~0.05~ 0.03
100.000.000 | 0.39 | ~0.26~ 0.23 | ~0.50~ 0.33 | 0.58 | ~0.45~ 0.37
134.217.728 (M) | 0.52 | ~0.34~ 0.29 | ~0.67~ 0.43 | 0.77 | ~0.59~ 0.52

</div>

</body>

</html>

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [x] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: dpnp/dpnp_iface_logic.py [hunk 1]
```diff
@@ -572,48 +546,109 @@ def any(a, /, axis=None, out=None, keepdims=False, *, where=True):
...
+    dpnp.check_supported_arrays_type(a, b, scalar_type=True)
+    dpnp.check_supported_arrays_type(
+        rtol, atol, scalar_type=True, all_scalars=True
     )
 
+    if dpnp.isscalar(b):
+        dt = dpnp.result_type(
+            b, dpnp.default_float_type(sycl_queue=a.sycl_queue)
+        )
+        b = dpnp.asarray(
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_logic.py:635]: That might cause an issue on Iris Xe:
```python
import dpnp

a = dpnp.array([1])
dpnp.isclose(a, 10**5)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[4], line 1
----> 1 dpnp.isclose(a, 10**5)

File ~/code/dpnp/dpnp/dpnp_iface_logic.py:635, in isclose(a, b, rtol, atol, equal_nan)
    631 if dpnp.isscalar(b):
    632     dt = dpnp.result_type(
    633         b, dpnp.default_float_type(sycl_queue=a.sycl_queue)
    634     )
--> 635     b = dpnp.asarray(
    636         b, dtype=dt, sycl_queue=a.sycl_queue, usm_type=a.usm_type
    637     )
    638 elif b.dtype.kind == "i":
    639     dt = dpnp.result_type(
    640         b, dpnp.default_float_type(sycl_queue=b.sycl_queue)
    641     )

File ~/code/dpnp/dpnp/dpnp_iface_arraycreation.py:517, in asarray(a, dtype, order, like, device, usm_type, sycl_queue)
    439 """
    440 Converts an input object into array.
    441
   (...)
    512
    513 """
    515 dpnp.check_limitations(like=like)
--> 517 return dpnp_container.asarray(
    518     a,
    519     dtype=dtype,
    520     order=order,
    521     device=device,
    522     usm_type=usm_type,
    523     sycl_queue=sycl_queue,
    524 )

File ~/code/dpnp/dpnp/dpnp_container.py:122, in asarray(x1, dtype, copy, order, device, usm_type, sycl_queue)
    116     x1_obj = x1
    118 sycl_queue_normalized = dpnp.get_normalized_queue_device(
    119     x1_obj, device=device, sycl_queue=sycl_queue
    120 )
--> 122 array_obj = dpt.asarray(
    123     x1_obj,
    124     dtype=dtype,
    125     copy=copy,
    126     order=order,
    127     usm_type=usm_type,
    128     sycl_queue=sycl_queue_normalized,
    129 )
    131 # return x1 if dpctl returns a zero copy of x1_obj
    132 if array_obj is x1_obj and isinstance(x1, dpnp_array):

File ~/miniforge3/envs/dpnp_dev/lib/python3.11/site-packages/dpctl/tensor/_ctors.py:648, in asarray(obj, dtype, device, copy, usm_type, sycl_queue, order)
    644     raise ValueError(
    645         f"Converting {type(obj)} to usm_ndarray requires a copy"
    646     )
    647 # obj is a scalar, create 0d array
--> 648 return _asarray_from_numpy_ndarray(
    649     np.asarray(obj, dtype=dtype),
    650     dtype=dtype,
    651     usm_type=usm_type,
    652     sycl_queue=sycl_queue,
    653     order="C",
    654 )

File ~/miniforge3/envs/dpnp_dev/lib/python3.11/site-packages/dpctl/tensor/_ctors.py:253, in _asarray_from_numpy_ndarray(ary, dtype, usm_type, sycl_queue, order)
    249     res = dpt.usm_ndarray(
    250         res.shape, dtype=res.dtype, buffer=res.usm_data, strides=new_strides
    251     )
    252 else:
--> 253     _ensure_native_dtype_device_support(dtype, copy_q.sycl_device)
    254     res = dpt.usm_ndarray(
    255         ary.shape,
    256         dtype=dtype,
   (...)
    259         buffer_ctor_kwargs={"queue": copy_q},
    260     )
    261 res[...] = ary

File ~/miniforge3/envs/dpnp_dev/lib/python3.11/site-packages/dpctl/tensor/_ctors.py:280, in _ensure_native_dtype_device_support(dtype, dev)
    266 """Check that dtype is natively supported by device.
    267
    268 Arg:
   (...)
    277         if device does not natively support this `dtype`.
    278 """
    279 if dtype in [dpt.float64, dpt.complex128] and not dev.has_aspect_fp64:
--> 280     raise ValueError(
    281         f"Device {dev.name} does not provide native support "
    282         "for double-precision floating point type."
    283     )
    284 if (
    285     dtype
    286     in [
   (...)
    289     and not dev.has_aspect_fp16
    290 ):
    291     raise ValueError(
    292         f"Device {dev.name} does not provide native support "
    293         "for half-precision floating point type."
    294     )

ValueError: Device Intel(R) Graphics [0x9a49] does not provide native support for double-precision floating point type.
```
I would propose to calculate dt here based on `a` also:
```python
        dt = dpnp.result_type(
            a, b, dpnp.default_float_type(sycl_queue=a.sycl_queue)
        )
```

File: dpnp/dpnp_iface_logic.py [hunk 2]
```diff
@@ -572,48 +546,109 @@ def any(a, /, axis=None, out=None, keepdims=False, *, where=True):
...
+            b, dtype=dt, sycl_queue=a.sycl_queue, usm_type=a.usm_type
+        )
+    elif b.dtype.kind == "i":
+        dt = dpnp.result_type(
+            b, dpnp.default_float_type(sycl_queue=b.sycl_queue)
+        )
+        b = dpnp.asarray(b, dtype=dt)
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_logic.py:642]:
```suggestion
        b = dpnp.astype(b, dtype=dt)
```

File: dpnp/dpnp_iface_logic.py [hunk 3]
```diff
@@ -572,48 +546,109 @@ def any(a, /, axis=None, out=None, keepdims=False, *, where=True):
...
 
+    if dpnp.isscalar(b):
+        dt = dpnp.result_type(
+            b, dpnp.default_float_type(sycl_queue=a.sycl_queue)
+        )
+        b = dpnp.asarray(
+            b, dtype=dt, sycl_queue=a.sycl_queue, usm_type=a.usm_type
+        )
+    elif b.dtype.kind == "i":
+        dt = dpnp.result_type(
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_logic.py:639]: Will it always result in `dpnp.default_float_type` or thee is the use case when the result differs from  `dpnp.default_float_type` ?

File: dpnp/dpnp_iface_logic.py [hunk 4]
```diff
@@ -572,48 +546,109 @@ def any(a, /, axis=None, out=None, keepdims=False, *, where=True):
...
+            b, dpnp.default_float_type(sycl_queue=b.sycl_queue)
+        )
+        b = dpnp.asarray(b, dtype=dt)
+
+    result = less_equal(dpnp.abs(a - b), atol + rtol * dpnp.abs(b)) & isfinite(
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_logic.py:644]: Should we use `out` keyword here to reduce temporary memory allocation? Will it bring the benefit?

File: tests/test_logic.py [hunk 5]
```diff
@@ -479,3 +475,22 @@ def test_infinity_sign_errors(func):
...
+
+
+@pytest.mark.parametrize("dtype", get_all_dtypes(no_bool=True, no_complex=True))
+@pytest.mark.parametrize(
+    "rtol", [1e-05, dpnp.array(1e-05), dpnp.full(10, 1e-05)]
+)
+@pytest.mark.parametrize(
+    "atol", [1e-08, dpnp.array(1e-08), dpnp.full(10, 1e-08)]
+)
+def test_isclose(dtype, rtol, atol):
```
Comment (antonwolfy) [inline tests/test_logic.py:487]: needs to uncomments tests in `test_sycl_queue.py` and `test_usm_type.py` for `isclose`

File: tests/third_party/cupy/logic_tests/test_comparison.py [hunk 6]
```diff
@@ -142,7 +142,6 @@ def test_allclose_infinite(self, xp, dtype):
 
     @testing.for_float_dtypes()
```
Comment (antonwolfy) [inline tests/third_party/cupy/logic_tests/test_comparison.py:143]: please update the file with the latest content

File: dpnp/dpnp_iface_logic.py (hunk 3 above)
Comment (npolina4) [reply dpnp/dpnp_iface_logic.py:639]: Changed to weak float dtype

File: dpnp/dpnp_iface_logic.py (hunk 4 above)
Comment (npolina4) [reply dpnp/dpnp_iface_logic.py:644]: Yes. Performance table has been updated.

File: dpnp/dpnp_iface_logic.py [hunk 7]
```diff
@@ -572,48 +546,115 @@ def any(a, /, axis=None, out=None, keepdims=False, *, where=True):
...
+def isclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
     """
     Returns a boolean array where two arrays are element-wise equal within
     a tolerance.
 
+    The tolerance values are positive, typically very small numbers. The
+    relative difference (`rtol` * abs(`b`)) and the absolute difference `atol`
+    are added together to compare against the absolute difference between `a`
+    and `b`.
+
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_logic.py:558]: That is applicable here as well:
```suggestion

    ``NaNs`` are treated as equal if they are in the same place and if
    ``equal_nan=True``. ``Infs`` are treated as equal if they are in the same
    place and of the same sign in both arrays.

```

File: dpnp/dpnp_iface_logic.py [hunk 8]
```diff
@@ -572,48 +546,115 @@ def any(a, /, axis=None, out=None, keepdims=False, *, where=True):
...
+    elif dpnp.issubdtype(b, dpnp.integer):
+        dt = dpnp.result_type(b, 1.0)
+        b = dpnp.astype(b, dtype=dt)
+
+    _b = dpnp.abs(b)
+    _b *= rtol
+    _b += atol
+    result = less_equal(dpnp.abs(a - b), _b)
+    result &= isfinite(b)
+    result |= a == b
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_logic.py:652]: It'd be great to add more clarity:
```suggestion
    # Firstly handle finite values:
    # result = absolute(a - b) <= atol + rtol * absolute(b)
    _b = dpnp.abs(b)
    _b *= rtol
    _b += atol
    result = less_equal(dpnp.abs(a - b), _b)

    # Handle "Inf" values: they are treated as equal if they are in the same place
    # and of the same sign in both arrays
    result &= isfinite(b)
    result |= a == b
```

**Review by antonwolfy — DISMISSED** [review]
Thank you @npolina4 No more comments from me.

File: tests/test_logic.py (hunk 5 above)
Comment (antonwolfy) [inline tests/test_logic.py:487]: The test is still failing.

**Review by antonwolfy — APPROVED** [review]

**Pull request merged**